### PR TITLE
feature: added collision detection functions

### LINKED ISSUE #82: Support Collision Detection API additions of Choreograph 4.8.0
Add support for new collision detection endpoints as detailed [here](https://docs.automata.tech/v/4.8.0/api/migration-guide/#collision-detection) & add new `collision` state to `robot_state.py` enum.

### PR DESCRIPTION
**Pull request by euanwm**
Fixes https://github.com/automata-tech/eva_python_sdk/issues/82

Changes:

- Added in the function collision_detection to allow enabling of collision detection including the sensitivity
- Added in the function acknowledge_collision as it is required post-collision
- IDE removed some whitespaces/formatting it wasn't happy with
- Added collision state to the RobotState

### CODE REVIEW

File: evasdk/Eva.py [hunk 1]
```diff
@@ -234,6 +234,14 @@ def control_reset_errors(self, wait_for_ready=True):
         self.__logger.debug('Eva.control_reset_errors called')
         return self.__http_client.control_reset_errors(wait_for_ready=wait_for_ready)
 
+    # Collision Detection
+    def collision_detection(self, on_off, setting):
```
Comment (LouisBrunner) [inline evasdk/Eva.py]: `set_collision_detection`, `configure_collision_detection` or similar would be better

File: evasdk/eva_http_client.py [hunk 2]
```diff
@@ -40,7 +42,6 @@ def __init__(self, host_ip, api_token, custom_logger=None, request_timeout=5, re
         if not 0 < renew_period < 30 * 60:
             raise ValueError('Session must be renewed before expiring (30 minutes)')
 
-
```
Comment (LouisBrunner) [inline evasdk/eva_http_client.py]: Minimum amount of changes please

File: evasdk/eva_http_client.py [hunk 3]
```diff
@@ -365,54 +342,70 @@ def control_go_to(self, joints, wait_for_ready=True, max_speed=None, time_sec=No
...
+            time.sleep(0.1)  # sleep for small period to avoid race condition between updating cache and reading state
             self.control_wait_for(RobotState.READY)
 
+    # COLLISION DETECTION
+    def collision_detection(self, on_off, setting):
+        """
+        Allows toggling on/off and setting the sensitivity of collision detection
+        """
+        if setting not in ['low', 'medium', 'high']:
+            eva_error('collision_detection error, no such sensitivity ' + setting)
```
Comment (LouisBrunner) [inline evasdk/eva_http_client.py]: We really shouldn't be using `eva_error` here but I guess that's what we have done so far... Maybe start switching to `raise ValueError`?

File: evasdk/eva_http_client.py [hunk 4]
```diff
@@ -365,54 +342,70 @@ def control_go_to(self, joints, wait_for_ready=True, max_speed=None, time_sec=No
...
+        if on_off not in [True, False]:
+            eva_error('collision_detection error, must be True/False')
```
Comment (LouisBrunner) [inline evasdk/eva_http_client.py]: Same here

File: evasdk/eva_http_client.py [hunk 5]
```diff
@@ -365,54 +342,70 @@ def control_go_to(self, joints, wait_for_ready=True, max_speed=None, time_sec=No
...
+        r = self.api_call_with_auth('POST', 'controls/collision_detection',
+                                    json.dumps({'enabled': on_off, 'sensitivity': setting}))
+        if r.status_code != 200:
+            eva_error('collision_detection error', r)
+
+    def acknowledge_collision(self):
```
Comment (LouisBrunner) [inline evasdk/eva_http_client.py]: Due to being a `controls/` endpoint, I think it should at least work like the `control_*` functions with the `wait_for_ready` et al
Maybe even being called with the `control_` prefix

**Review by LouisBrunner — CHANGES_REQUESTED** [review]

File: evasdk/Eva.py (hunk 1 above)
Comment (tafautomata) [inline evasdk/Eva.py]: `enabled` might be more descriptive than `on_off` as func. parameter name, and maybe `sensitivity` for `setting`

File: evasdk/eva_http_client.py (hunk 2 above)
Comment (euanwm) [reply evasdk/eva_http_client.py]: PyCharm went rogue. Will try and rollback this.

File: evasdk/eva_http_client.py [hunk 6]
```diff
@@ -413,6 +415,30 @@ def control_reset_errors(self, wait_for_ready=True):
...
+        Allows toggling on/off and setting the sensitivity of collision detection
+        """
+        if sensitivity not in ['low', 'medium', 'high']:
+            raise ValueError('collision_detection error, no such sensitivity ' + sensitivity)
+        if enabled not in [True, False]:
+            raise ValueError('collision_detection error, must be True/False')
+        r = self.api_call_with_auth('POST', 'controls/collision_detection',
+                                    json.dumps({'enabled': enabled, 'sensitivity': sensitivity}))
+        if r.status_code != 200:
+            raise ValueError('control_collision_detection error')
```
Comment (LouisBrunner) [inline evasdk/eva_http_client.py]: Sorry, misunderstanding: request related exceptions should stay `eva_error`, arguments errors should be `ValueError`, so `r.status_code` in both functions should stay `eva_error`

File: evasdk/eva_http_client.py [hunk 7]
```diff
@@ -413,6 +415,30 @@ def control_reset_errors(self, wait_for_ready=True):
...
+
+    def control_acknowledge_collision(self, wait_for_ready=True):
+        """
+        When a collision is encountered, it must be acknowledged before the robot can be reset
+        """
+        r = self.api_call_with_auth('POST', 'controls/acknowledge_collision')
+        if r.status_code != 204:
+            raise ValueError('control_acknowledge_collision error')
+        elif wait_for_ready:
+            time.sleep(0.1)
```
Comment (LouisBrunner) [inline evasdk/eva_http_client.py]: Add the same comment?

**Review by LouisBrunner — CHANGES_REQUESTED** [review]

File: evasdk/eva_http_client.py [hunk 8]
```diff
@@ -413,6 +415,30 @@ def control_reset_errors(self, wait_for_ready=True):
...
+        """
+        Allows toggling on/off and setting the sensitivity of collision detection
+        """
+        if sensitivity not in ['low', 'medium', 'high']:
+            raise ValueError('collision_detection error, no such sensitivity ' + sensitivity)
+        if enabled not in [True, False]:
+            raise ValueError('collision_detection error, must be True/False')
+        r = self.api_call_with_auth('POST', 'controls/collision_detection',
+                                    json.dumps({'enabled': enabled, 'sensitivity': sensitivity}))
+        if r.status_code != 200:
```
Comment (walker-j) [inline evasdk/eva_http_client.py]:
```suggestion
        if r.status_code != 204:
```

**Review by walker-j — DISMISSED** [review]
Tested against 4.9.0, lgtm

**Comment by walker-j** [comment]
Closing this PR as these changed were included in PR #84

**Pull request closed without merging**